nntrainer/src/tensor.cpp

```diff
@@ -592,143 +592,118 @@ Tensor Tensor::sum(int axis) const {
   return ret;
 }
 
-/**
- * If the dim.batch() size of m is one, the it is reused for
- * every calculation along with dim.batch().
- * TODO: support dim.channel() > 1.
- */
-Tensor Tensor::dot(Tensor const &m) const {
-  if (dim.width() != m.dim.height()) {
-    throw std::runtime_error("Error dim.width() != m.dim.height()");
+Tensor Tensor::dot(Tensor const &m, bool trans, bool trans_m) const {
+  if (m.dim.rank() > 2) {
+    throw std::runtime_error("Error: support only for rank of dot matrix <= 2");
   }
 
-  if (dim.channel() != 1 || m.dim.channel() != 1) {
-    throw std::runtime_error("Error channel() != 1");
-  }
+  unsigned int dim1 = batch() * channel() * height();
+  unsigned int dim2 = width();
+  unsigned int mdim1 = m.batch() * m.channel() * m.height();
+  unsigned int mdim2 = m.width();
+  Tensor result;
 
-  int mwidth = m.dim.width();
-  Tensor result(dim.batch(), 1, dim.height(), mwidth);
+  unsigned int M, N, K, lda, ldb, ldc;
+
+  if (!trans && !trans_m) {
+    if (dim2 != mdim1)
+      throw std::runtime_error(
+        "Error: imcompatible dimensions for dot product");
```

Comment:
```suggestion
        "Error: incompatible dimensions for dot product");
```

And for others below

Reply:
done.